lib/IDE/Refactoring.cpp

```diff
@@ -5245,44 +5253,51 @@ class AsyncConverter : private SourceEntityWalker {
     OS << tok::l_brace;
   }
 
-  void addCatchBody(const ParamDecl *ErrParam,
-                    const ClassifiedBlock &ErrorBlock) {
-    convertNodes(ErrorBlock.nodes());
-    OS << "\n" << tok::r_brace;
-  }
+  void preparePlaceholdersAndUnwraps(AsyncHandlerDesc HandlerDesc,
+                                     ArrayRef<const ParamDecl *> SuccessParams,
+                                     const ParamDecl *ErrParam, bool Success) {
+    if (HandlerDesc.Type == HandlerType::RESULT) {
+      // Any uses of the result parameter in the current body (that
+      // isn't replaced) are invalid, so replace them with a placeholder
+      Placeholders.insert(ErrParam);
+      return;
+    }
 
-  void prepareNamesForBody(const AsyncHandlerDesc &HandlerDesc,
-                           ArrayRef<const ParamDecl *> CurrentParams,
-                           ArrayRef<const ParamDecl *> OtherParams) {
-    switch (HandlerDesc.Type) {
-    case HandlerType::PARAMS:
-      for (auto *Param : CurrentParams) {
-        auto Ty = Param->getType();
-        if (Ty->getOptionalObjectType()) {
-          Unwraps.insert(Param);
-          Placeholders.insert(Param);
+    assert(HandlerDesc.Type == HandlerType::PARAMS &&
+           "Unhandled handler type");
+
+    if (!Success) {
+      if (ErrParam) {
+        if (ErrParam->getType()->getOptionalObjectType()) {
+          Placeholders.insert(ErrParam);
+          Unwraps.insert(ErrParam);
         }
-        // Void parameters get omitted where possible, so turn any reference
-        // into a placeholder, as its usage is unlikely what the user wants.
-        if (HandlerDesc.getSuccessParamAsyncReturnType(Ty)->isVoid())
-          Placeholders.insert(Param);
```

Comment:
Can we keep this logic? I realise now that the test case I added for it was subsumed by the `Ty->getOptionalObjectType()` check, but this test case should demonstrate it:

```
func foo(_ completion: (Void, Error?) -> Void) {}
foo { v, err in
  print(v)
}
```

When re-writing the call to `foo` we should turn the reference to `v` into a placeholder. It's a silly example, I know, but I'd rather not make it invalid without introducing a placeholder.

Reply:
Whoops! Thanks, I had this change sitting in a branch and mostly just fixed up the conflicts - missed this part. I'll add that test case.